specs/sequencer-relayer.md

```diff
@@ -0,0 +1,245 @@
+# Astria Sequencer Relayer
+
+The Astria Sequencer Relayer (Relayer), is a stateless service which reads new
+blocks from the sequencer, publishes them to the gossipnet for rollup execution,
+and writes batches of sequencer blocks to DA.
+
+It is run as a sidecar to a sequencer validator node, and each instance only
+relays/writes data from blocks it proposes.
+
+## Interfaces
+
+### API
+
+Relayer offers basic status/health check endpoints via an HTTP server with the
+following endpoints:
+
+- `readyz`
+  - path: `/readyz`
+  - input: none
+  - output either:
+    - status: 200, `ok`
+    - status: 503, `not ready`
+- `healthz`
+  - path: `/healthz`
+  - input: none
+  - output either:
+    - status: 200, `ok`
+    - status: 504, `degraded`
+- `status`
+  - path: `/status`
+  - input: none
+  - output:
+    - json:
+      - `data_availability_connected`
+        - boolean status of DA connection,
+      - `sequencer_connected`
+        - boolean status of sequencer connection
+      - `current_sequencer_height`
+        - integer height of last seen sequencer block
+      - `current_data_availability_height`
+        - integer height of last known DA block height
+      - `number_of_subscribed_peers`
+        - integer number of subscribed gossipnet members
+
+### Sequencer
+
+Relayer fetches new blocks from the CometBFT consensus node of the sequencer
+using [CometBFT's Block RPC](https://docs.cometbft.com/v0.37/spec/rpc/#block)
+with the `height` set to null to grab the latest block. This returns the latest
+proposed block, which is not fully finalized by the sequencer network yet.
+
+### P2P/Gossipnet
+
+> Note: The P2P/gossipnet may be removed/replaced by an alternate model in the
+> future. It's known that this is heavyweight and is propogating data to all
+> rollups which they are not all interested in.
+
+After receiving a block from the sequencer, if it is signed by the relayer's
+validator key, it is converted into the `SequencerBlockData` shape
+([link](https://github.com/astriaorg/astria/blob/7ebb743ed6f1d9eef69372f2cbb4ab9cbe2668b3/crates/astria-sequencer-types/src/sequencer_block_data.rs#L39-L46)).
+This object is then pushed to the libp2p "gossipnet" network, for execution by
+rollups. This contains information for execution by all rollups.
+
+```rust
+/// `SequencerBlockData` represents a sequencer block's data
+/// to be submitted to the DA layer.
+pub struct SequencerBlockData {
+    pub(crate) block_hash: Hash,
+    pub(crate) header: Header,
+    /// This field should be set for every block with height > 1.
+    pub(crate) last_commit: Option<Commit>,
+    /// namespace -> rollup txs
+    pub(crate) rollup_txs: HashMap<Namespace, Vec<Vec<u8>>>,
+}
+```
+
+### Data Availability
+
+Relayer also writes data to Celestia DA. Since Astria block times can be much
+faster than Celestia block times, we create a queue of yet to be written blocks.
```

Comment:
maybe explicitly say "we make 1 batch write of data per celestia block, which consists of many sequencer blocks"?